### DISCUSSION TITLE
How to annotate a function to return an instance of a passed-in type expression?

### DISCUSSION THREAD

**Discussion by ToBeReplaced** [microsoft/pyright#7180] [Q&A] [answered]
There are a few ongoing discussions about how to handle the new `TypeAliasType`, and there are recent changes to pyright (starting in 1.1.346) affecting how `UnionType` and `Literal` behave with respect to type variables.

See: 
https://discuss.python.org/t/run-time-behaviour-of-typealiastype/43774
https://github.com/microsoft/pyright/issues/6895
https://github.com/microsoft/pyright/issues/7110

How can I define a function such that accepts a type expression, and returns an instance that satisfies that expression?

For example, for anything that is a subtype of `type`:

```python
def spam[T](type_: type[T]) -> T:
    ...
```

Consider now that I want to allow the user to pass in `UnionType`. For example, `spam(int | str)`. This is disallowed, as `type[int] | type[str]` cannot be assigned to `type[T@spam]`.

Prior to 1.1.346, this was allowed.

We can make a `UnionType` permissable, but there is no means of using it with a type variable.

For example:

```python
def eggs(type_: type[Any] | UnionType) -> Any:
    ...
```

If we include a type variable...

```python
def eggs[T](type_: type[T] | UnionType) -> T:
    ...
```

... then `eggs(int | str)` has type Unknown. We would like for it to become `int | str`. Is there any way to achieve this pattern?

It has been suggested that if you want to pass `UnionType`, you could instead pass a tuple of types, but this hits the same problem in a different flavor; How can you then join the types into a union?

```python
@typing.overload
def ham[*Ts](type_: tuple[*Ts]) -> Any:
    ...


@typing.overload
def ham[T](type_: type[T]) -> T:
    ...


def ham[T, *Ts](type_: type[T] | tuple[*Ts]) -> Any:
    ...

```

Notice that we now also have no way of indicating that the tuple must contain types, as we cannot put a bound on a variadic type parameter.

Interestingly, this type-checks and makes `ham(int | str)` have type `int | str`, but I suspect this is a bug.

Similarly, this is not what you want:

```python
def ham[T](type_: type[T] | tuple[type[T], ...]) -> Any:
    ...
```

As far as I can tell, the changes to `UnionType` make it impossible to type-check code where the caller wants to pass in one-or-more types and receive a result that is an instance of one of those types.

If we look at the problem from the lens of `TypeAliasType`, all of the same issues appear, in a slightly different flavor:

```python

def alias_attempt[T: TypeAliasType](type_: T) -> T.__value__:  # Cannot do this...
    ...

```

How do we go from `TypeAliasType` to saying the function will return an instance that would satisfy the `TypeAliasType`?

If we were to be able to do that, how could we combine that with `UnionType` and `type`?

As a heavy user of this kind of "dispatching" code, I am currently locked to 1.1.345. Is there any pattern I could use to begin upgrading?

**Comment by erictraut** [comment]
This is probably not the right forum to be discussing this. If you would like to clarify your use case and make a pitch for how the Python type system should be extended to support your use case, the [typing forum](https://discuss.python.org/c/typing/32) would be the best place for that.

There has been discussion in the past of adding support for `typing.TypeForm` that would accept any type form that is accepted within a type annotation. It's possible that this is what you're looking for.

**Comment by ToBeReplaced** [comment]
Yes, `TypeForm` is an interesting concept.

I posted this discussion here because the use case in question was checked with pyright in 1.1.345, and now the only known approach is to lose information about the return type, making it easier to write incorrect programs.

Since upgrading pyright requires me to make my annotations _less strict_, I thought I should see if pyright has any mechanism of recovering those details.

Let's consider only the `UnionType` use case.

```python3
def spam[T](type_: type[T]) -> T:
    ...
```

Old behavior maps `spam(int) -> int` and `spam(int | str) -> int | str`. New behavior disallows passing the `UnionType`.

So I should write the following and lose the type-narrowing at all call sites, correct?
```python3
def spam(type_: type[Any] | UnionType) -> Any:
    ...
```

**Reply by erictraut** [reply, marked as answer]
Pyright's old behavior didn't comply with the Python typing standard. The recent changes were part of my effort to tighten up pyright's [compliance](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html). Sorry that this is causing you short-term pain, but I think it's in everyone's best interest in the long run.

**Reply by ToBeReplaced** [reply]
Understood.

Hopefully `TypeForm` or something like it arrives in the next few years. If that is too long for me to wait, I should get involved :wink:. There is no one opposing, just work to be done.

Thank you for all of the work you do both with pyright and advancing Python's type-system.
